test/unittest/unittest_nntrainer_tensor_pool.cpp

```diff
@@ -0,0 +1,364 @@
+ */
+TEST(TensorPool, finalize_02_p) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1, *t2;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {0},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc1"));
+  EXPECT_NE(t1, nullptr);
+
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_NO_THROW(
+    t2 = pool.requestTensor(nntrainer::TensorDim({1}), {1},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc2"));
+  EXPECT_NE(t2, nullptr);
+
+  EXPECT_FALSE(t2->isAllocated());
+
+  EXPECT_NE(t1, t2);
+
+  EXPECT_NO_THROW(pool.finalize(nntrainer::BasicPlanner(), 0, 2));
+  EXPECT_EQ(pool.minMemoryRequirement(), t1->bytes() + t2->bytes());
+
+  EXPECT_FALSE(t1->isAllocated());
+  EXPECT_FALSE(t2->isAllocated());
+}
+
+/**
+ * @brief max lifespan tensors
+ */
+TEST(TensorPool, finalize_03_p) {
+  nntrainer::TensorPool pool;
+
+  EXPECT_NO_THROW(pool.finalize(nntrainer::BasicPlanner(), 0, 2));
+}
+
+/**
+ * @brief allocate
+ */
+TEST(TensorPool, allocate_deallocate_01_p) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1, *t2;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({1}), {0},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc1"));
+  EXPECT_NE(t1, nullptr);
+  EXPECT_FALSE(t1->isAllocated());
+
+  EXPECT_NO_THROW(
+    t2 = pool.requestTensor(nntrainer::TensorDim({1}), {1},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc2"));
+  EXPECT_NE(t2, nullptr);
+  EXPECT_FALSE(t2->isAllocated());
+
+  EXPECT_NE(t1, t2);
+
+  EXPECT_NO_THROW(pool.finalize(nntrainer::BasicPlanner(), 0, 2));
+
+  EXPECT_NO_THROW(pool.allocate());
+  EXPECT_TRUE(t1->isAllocated());
+  EXPECT_TRUE(t2->isAllocated());
+
+  EXPECT_NO_THROW(pool.deallocate());
+  EXPECT_FALSE(t1->isAllocated());
+  EXPECT_FALSE(t2->isAllocated());
+}
+
+/**
+ * @brief allocate
+ */
+TEST(TensorPool, allocate_deallocate_02_n) {
+  nntrainer::TensorPool pool;
+
+  EXPECT_THROW(pool.allocate(), std::runtime_error);
+
+  EXPECT_NO_THROW(pool.deallocate());
+}
+
+/**
+ * @brief validate memory full overlap
+ */
+TEST(TensorPool, validate_memory) {
+  nntrainer::TensorPool pool;
+  nntrainer::Tensor *t1, *t2;
+
+  EXPECT_NO_THROW(
+    t1 = pool.requestTensor(nntrainer::TensorDim({100}), {0},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc1"));
+
+  EXPECT_NO_THROW(
+    t2 = pool.requestTensor(nntrainer::TensorDim({100}), {1},
+                            nntrainer::TensorLifespan::MAX_LIFESPAN, "abc2"));
+
+  EXPECT_NO_THROW(pool.finalize(nntrainer::BasicPlanner(), 0, 2));
+  EXPECT_NO_THROW(pool.allocate());
+
+  nntrainer::Tensor g1 = nntrainer::Tensor(nntrainer::TensorDim({100}));
+  g1.setRandNormal();
+  nntrainer::Tensor g2 = nntrainer::Tensor(nntrainer::TensorDim({100}));
+  g2.setRandNormal();
+
+  t1->copy(g1);
+  t2->copy(g2);
+
+  EXPECT_EQ(*t1, g1);
+  EXPECT_EQ(*t2, g2);
```

Comment:
Wonder what this lines try to test? is this trying to check t1 and t2 does not clash?
If that's the case, I recommend to use Tensor::fill. copy creates a new tensor if it has empty data.

Reply:
These checks ensure that the memory of tensors t1 and t2 are not overlapped.